src/credentials/CHIPCert.cpp

```diff
@@ -708,20 +709,122 @@ CHIP_ERROR DetermineCertType(ChipCertificateData & cert)
 
 bool ChipDN::IsEqual(const ChipDN & other) const
 {
-    if (mAttrOID == kOID_Unknown || mAttrOID == kOID_NotSpecified || mAttrOID != other.mAttrOID)
-    {
+    // Same number of attributes?
+    if (mCount != other.mCount)
         return false;
+
+    // kOID_AttributeType_ChipNodeId has priority and is located at index 0, any other are at higher indices.
+    if (mAttrOID[0] == chip::ASN1::kOID_AttributeType_ChipNodeId)
+        return other.mAttrOID[0] == chip::ASN1::kOID_AttributeType_ChipNodeId &&
+            mAttrValue[0].mChipId == other.mAttrValue[0].mChipId;
```

Comment:
CHIP OpCert for a node will have `Subject: 1.3.6.1.4.1.37244.1.1 = DEDEDEDE00000001, 1.3.6.1.4.1.37244.1.5 = FAB000000000001D`. So we will have > 1 RDN.

Reply:
@hawk248, the question was how do we compare two DNs. In alignment with our discussion, my goal was to give kOID_AttributeType_ChipNodeId a priority; that is for two DNs to be equal is enough to have the same Node Id's. Such rule seems to be not very intuitive. Please confirm I captured the requirement properly.

Reply:
CHIP OpCert for a node will have _at least_ 2 RDNs.  It will always have a fabric id and node id.  After https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/2263 merges it could have up to 4 RDNs, not counting any non-CHIP bits.  As far as how two DNs are compared, the spec pretty clearly says (in section "6.4.8.1. X.501 Distinguished Names"):

> Two distinguished names DN1 and DN2 match if they have the same number of RDNs, for each RDN in DN1 there is a matching RDN in DN2, and the matching RDNs appear in the same order in both DNs.

And under "6.4.8.3. CHIP DN Encoding Rules":

> The RDNs SHALL appear in the same order in the CHIP certificate and in the corresponding X.509 certificates.
> ...
> The order of the RDNs can be issuer-specific and is not enforced by CHIP specifications.

If I read all that right, that means no special-casing the node id rdn for either equality or ordering purposes, no?

Reply:
Yes, that's how I read it.

I'm missing something though.  I don't see anything that calls out the appearance of multiple `chip-node-id` attributes or multiple `chip-fabric-id` attributes as an error.  The reason x509 RDN ordering matters is that attributes can be repeated.  The DN is not a map; it is an ordered list of key=value pair RDNs.

What happens if you mDNS advertise for fabric a and node ID x, but offer a certificate with RDNs for fabrics a, b, and c and node IDs x, y and z?

Reply:
This part of the spec:

> There SHALL be only one CHIP-specific certificate type DN attribute from the table Table 51, “CHIP Certificate Types” encoded in the subject DN of the CHIP certificate.

means you can only have one chip-node-id RDN, though I agree it's not immediately obvious that it means that.  But you're right that nothing prevents encoding more than one chip-fabric-id.  I filed https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/2353 on making this stuff more readable and correct....